src/parakeet/ProE/internal/Parser.cpp

```diff
@@ -0,0 +1,348 @@
+    {
+        //We are generating the timestamp first, so that we can get the most accurate time.
+        generateTimestamp();
+
+        parseNumPointsInThisPartialSector();
+        parseNumPointsInSector();
+        parseSectorDataOffset();
+        parseStartAngle();
+        parseEndAngle();
+        parseSensorProperties();
+        parseTimestamp();
+        parseDeviceNumber();
+
+        parsePoints();
+
+        parseChecksum();
+    }
+
+    bool Parser::doesChecksumMatch()
+    {
+        uint16_t newChecksum = 0;
+
+        newChecksum += currentLidarMessage->numPoints;
+        newChecksum += currentLidarMessage->numPointsInSector;
+        newChecksum += currentLidarMessage->sectorDataOffset;
+
+        newChecksum += currentLidarMessage->startAngle >> 16;
+        newChecksum += currentLidarMessage->startAngle & 0xFFFF;
+
+        newChecksum += currentLidarMessage->endAngle >> 16;
+        newChecksum += currentLidarMessage->endAngle & 0xFFFF;
+
+        newChecksum += currentLidarMessage->sensorPropertyFlags.value >> 16;
+        newChecksum += currentLidarMessage->sensorPropertyFlags.value & 0xFFFF;
+
+        newChecksum += currentLidarMessage->timestamp >> 16;
+        newChecksum += currentLidarMessage->timestamp & 0xFFFF;
+
+        newChecksum += currentLidarMessage->deviceNumber >> 16;
+        newChecksum += currentLidarMessage->deviceNumber & 0xFFFF;
+
+        for(LidarPoint lidarPoint : currentLidarMessage->lidarPoints)
+        {
+            newChecksum += lidarPoint.distance;
+            newChecksum += lidarPoint.relativeStartAngle;
+            newChecksum += lidarPoint.intensity;
+        }
+
+        return newChecksum == currentLidarMessage->checksum;
+    }
+
+    int Parser::getTotalPointsOfAllPartialScans()
+    {
+        int totalPoints = 0;
+        for (auto sector : partialSectorScanDataList)
+        {
+            totalPoints += sector->numPoints;
+        }
+
+        return totalPoints;
+    }
+
+    bool Parser::isScanComplete()
+    {
+        return getTotalPointsOfAllPartialScans() == currentLidarMessage->numPointsInSector;
+    }
+
+    bool Parser::isScanCorrupt()
+    {
+        if (getTotalPointsOfAllPartialScans() > currentLidarMessage->numPointsInSector)
+        {
+            return true;
+        }
+
+        if (!doesChecksumMatch())
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+    void Parser::createAndPublishCompleteScan()
+    {
+        if (partialSectorScanDataList.size() <= 0)
+        {
+            return;
+        }
+
+        auto firstMessage = partialSectorScanDataList[0];
+
+        //We will not ship the information from the first revolution, as it will not have a valid timestamp.
+        if (!firstMessage->generatedTimestamp.validTimestamp)
+        {
+            partialSectorScanDataList.clear();
+            return;
+        }
+
+        CompleteLidarMessage* lidarMessage = new CompleteLidarMessage();
```

Comment:
Do we need a pointer here? Seems like we can use a non-pointer instance here and use the same approach as I suggested for the ScanData and the onScanDataReceived callback.